pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,336 @@
+		LabelSelector: labels.SelectorFromSet(labels.Set{fleetv1beta1.CRPTrackingLabel: crp.Name}),
+	}
+	// TO-DO (chenyu1): this is a very expensive op; explore options for optimization.
+	if err := s.uncachedReader.List(ctx, bindingList, listOptions); err != nil {
+		klog.ErrorS(err, "Failed to list all bindings", "ClusterResourcePlacement", crpRef)
+		return controller.NewAPIServerError(false, err)
+	}
+
+	// Remove the scheduler cleanup finalizer from all the bindings, and delete them.
+	//
+	// Note that once a CRP has been marked for deletion, it will no longer enter the scheduling cycle,
+	// so any cleanup finalizer has to be removed here.
+	//
+	// Also note that for deleted CRPs, derived bindings are deleted right away by the scheduler;
+	// the scheduler no longer marks them as deleting and waits for another controller to actually
+	// run the deletion.
+	for idx := range bindingList.Items {
+		binding := bindingList.Items[idx]
+		// Delete the binding if it has not been marked for deletion yet.
+		if binding.DeletionTimestamp == nil {
+			if err := s.client.Delete(ctx, &binding); err != nil {
+				klog.ErrorS(err, "Failed to delete binding", "clusterResourceBinding", klog.KObj(&binding))
+				return controller.NewAPIServerError(false, err)
+			}
+		}
+
+		// Note that the scheduler will not add any cleanup finalizer to a binding.
+	}
+
+	// All bindings have been deleted; remove the scheduler cleanup finalizer from the CRP.
+	controllerutil.RemoveFinalizer(crp, fleetv1beta1.SchedulerCRPCleanupFinalizer)
+	if err := s.client.Update(ctx, crp); err != nil {
+		klog.ErrorS(err, "Failed to remove scheduler cleanup finalizer from cluster resource placement", "clusterResourcePlacement", crpRef)
+		return controller.NewAPIServerError(false, err)
+	}
+
+	return nil
+}
+
+// lookUpLatestPolicySnapshot returns the latest (i.e., active) policy snapshot associated with
+// a CRP.
+func (s *Scheduler) lookUpLatestPolicySnapshot(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) (*fleetv1beta1.ClusterSchedulingPolicySnapshot, error) {
+	crpRef := klog.KObj(crp)
+
+	// Find out the latest policy snapshot associated with the CRP.
+	policySnapshotList := &fleetv1beta1.ClusterSchedulingPolicySnapshotList{}
+	labelSelector := labels.SelectorFromSet(labels.Set{
+		fleetv1beta1.CRPTrackingLabel:      crp.Name,
+		fleetv1beta1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+	})
+	listOptions := &client.ListOptions{LabelSelector: labelSelector}
+	// The scheduler lists with a cached client; this does not have any consistency concern as sources
+	// will always trigger the scheduler when a new policy snapshot is created.
+	if err := s.client.List(ctx, policySnapshotList, listOptions); err != nil {
+		klog.ErrorS(err, "Failed to list policy snapshots of a cluster resource placement", "clusterResourcePlacement", crpRef)
+		return nil, controller.NewAPIServerError(true, err)
+	}
+
+	switch {
+	case len(policySnapshotList.Items) == 0:
```

Comment:
The scheduler may be trigger by the cluster change. At that time, CRP is just removing the latest label from the old snapshot and has not created a new one.

Reply:
The logic remains the same, right?

Reply:
The logic is the same. Need to update the comments & returned the expectedBehavior error.